File: My 120+ Solutions!/Minimum Path Sum/Minimum_Path_Sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0){
            return 0;
        }
        int [][] dp = new int [grid.length][grid[0].length];
        for (int row = 0; row < dp.length; row++){
            for (int col = 0; col < dp[row].length; col++){
                dp[row][col] = grid[row][col];
                if (row > 0 && col > 0){
                    dp[row][col] += Math.min(dp[row-1][col],dp[row][col-1]);
                }
                else if (row > 0){
                    dp[row][col] += dp[row-1][col];
                }
                else if (col > 0){
                    dp[row][col] += dp[row][col-1];
                }
            }
        }
        return dp[dp.length-1][dp[0].length-1];
    }
}

File: Coding Problems & My Solutions/Minimum Path Sum/Minimum_Path_Sum.swift
class Solution {
    func minPathSum(_ grid: [[Int]]) -> Int {
        var dp = [[Int]](repeating: [Int](repeating: 0, count: grid[0].count), count: grid.count)
        for i in 0..<grid.count {
            for j in 0..<grid[i].count {
                dp[i][j] += grid[i][j]
                var minFromAbove = Int.max
                var minFromLeft = Int.max
                if i - 1 >= 0 {
                    minFromAbove = dp[i - 1][j]
                }
                if j - 1 >= 0 {
                    minFromLeft = dp[i][j - 1]
                }
                var minOfAboveOrBelow = min(minFromAbove, minFromLeft)
                if minOfAboveOrBelow != Int.max {
                    dp[i][j] += minOfAboveOrBelow
                }
            }
        }
        return dp[dp.count - 1][dp[0].count - 1]
    }
}